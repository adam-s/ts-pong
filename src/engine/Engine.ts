/* eslint-disable @typescript-eslint/no-empty-function */
import { Loadable, Loader } from '../loader';
import { timestamp, PartialBy } from '../utils';

export interface Game {
  init?: () => void;
  width: number;
  height: number;
  start: (numPlayers: number) => void;
  stop: () => void;
  update: (dt: number) => void;
  draw: (ctx: CanvasRenderingContext2D) => void;
  onkeyup?: (ev: KeyboardEvent) => void;
  onkeydown?: (ev: KeyboardEvent) => void;
}

export interface IGameOptions {
  fps?: number;
  stats?: boolean;
  width?: number;
  height?: number;
  canvasElementId?: string;
}

const defaultOptions: IGameOptions = {
  fps: 60,
  stats: true,
  width: 640,
  height: 480,
  canvasElementId: 'game',
};

export class Engine {
  private _options: IGameOptions;

  private _timer: ReturnType<typeof setInterval>;

  private _lastFrame: number;

  private _fps: number;

  private _interval: number;

  private _canvasElementId: string;

  private _width: number;

  private _height: number;

  private _front: HTMLCanvasElement;

  private _back: HTMLCanvasElement;

  private _front2d: CanvasRenderingContext2D;

  private _back2d: CanvasRenderingContext2D;

  private _loader: Loader;

  private _isLoading = false;

  private _hasStarted = false;

  private _stats = {
    count: 0,
    fps: 0,
    update: 0,
    draw: 0,
    frame: 0, // update + draw
  };

  constructor(
    private game: Game,
    options?: PartialBy<IGameOptions, 'canvasElementId'>
  ) {
    // For some inspiration how to include options to a TypeScript class
    // https://github.com/microsoft/vscode/blob/main/src/vs/editor/contrib/zoneWidget/zoneWidget.ts
    this._options = { ...defaultOptions, ...options };

    this._fps = this._options.fps;
    this._interval = 1000.0 / this._options.fps;

    // Set or create the front canvas element
    if (this._options.canvasElementId) {
      this._front = <HTMLCanvasElement>(
        document.getElementById(this._options.canvasElementId)
      );
    } else {
      this._front = <HTMLCanvasElement>document.createElement('canvas');
    }

    // Create a back canvas element
    this._back = <HTMLCanvasElement>document.createElement('canvas');

    this._width = this.game.width;
    this._height = this.game.height;

    this._front.width = this._back.width = this._width;
    this._front.height = this._back.height = this._height;

    this._front2d = this._front.getContext('2d');
    this._back2d = this._back.getContext('2d');

    this._loader = new Loader();

    this.addEvents();
  }

  async start(loader?: Loader): Promise<void> {
    // if (!this._compatible) {
    //   return Promise.reject('This game is incompatible with your browser');
    // }
    let loadingComplete: Promise<void>;
    if (loader) {
      this._loader = loader;
      loadingComplete = this.load(this._loader);
    } else {
      loadingComplete = Promise.resolve();
    }
    await loadingComplete;

    // Give an options to call and async init in the case the game requires loading resources
    this.game.init?.();

    this._lastFrame = timestamp();
    this._timer = setInterval(() => {
      this.loop();
    }, this._interval);
  }

  public load(loader: Loadable<unknown>): Promise<void> {
    const complete = new Promise<void>((resolve) => {
      this._isLoading = true;
      loader.load().then(() => {
        this._isLoading = false;
        resolve();
      });
    });

    return complete;
  }

  stop(): void {
    clearInterval(this._timer);
  }

  loop(): void {
    const start = timestamp();
    this.update((start - this._lastFrame) / 1000.0); // send dt as seconds

    const middle = timestamp();
    this.draw();

    const end = timestamp();
    this.updateStats(middle - start, end - middle);
    this._lastFrame = start;
  }

  update(dt: number): void {
    // might as well let the engine watch key press and inform the game on every update rather than pass through
    // with an eventing
    this.game.update(dt);
  }

  draw(): void {
    this._back2d.clearRect(0, 0, this._width, this._height);
    this.game.draw(this._back2d);
    this.drawStats(this._back2d);
    this._front2d.clearRect(0, 0, this._width, this._height);
    this._front2d.drawImage(this._back, 0, 0);
  }

  resetStats() {
    this._stats = {
      count: 0,
      fps: 0,
      update: 0,
      draw: 0,
      frame: 0, // update + draw
    };
  }

  updateStats(update: number, draw: number) {
    if (this._options.stats) {
      this._stats.update = Math.max(1, update);
      this._stats.draw = Math.max(1, draw);
      this._stats.frame = this._stats.update + this._stats.draw;
      this._stats.count =
        this._stats.count == this._fps ? 0 : this._stats.count + 1;
      this._stats.fps = Math.min(this._fps, 1000 / this._stats.frame);
    }
  }

  drawStats(ctx: CanvasRenderingContext2D) {
    if (this._options.stats) {
      ctx.fillText(
        'frame: ' + this._stats.count,
        this._width - 100,
        this._height - 60
      );
      ctx.fillText(
        'fps: ' + this._stats.fps,
        this._width - 100,
        this._height - 50
      );
      ctx.fillText(
        'update: ' + this._stats.update + 'ms',
        this._width - 100,
        this._height - 40
      );
      ctx.fillText(
        'draw: ' + this._stats.draw + 'ms',
        this._width - 100,
        this._height - 30
      );
    }
  }

  addEvents() {
    document.addEventListener('keyup', this.onkeyup.bind(this), false);
    document.addEventListener('keydown', this.onkeydown.bind(this), false);
  }

  onkeyup(ev: KeyboardEvent) {
    if (this.game.onkeyup) this.game.onkeyup(ev);
  }

  onkeydown(ev: KeyboardEvent) {
    if (this.game.onkeydown) this.game.onkeydown(ev);
  }

  hideCurser() {
    this._front.style.cursor = 'none';
  }

  showCurser() {
    this._front.style.cursor = 'auto';
  }

  alert(msg: string) {
    this.stop(); // alert blocks thread, so need to stop game loop in order to avoid sending huge dt values to next update
    const result = window.alert(msg);
    this.start();
    return result;
  }

  confirm(msg: string) {
    this.stop(); // alert blocks thread, so need to stop game loop in order to avoid sending huge dt values to next update
    const result = window.confirm(msg);
    this.start();
    return result;
  }

  showStats(show: boolean) {
    this._options.stats = show;
  }

  static random(min: number, max: number) {
    return min + Math.random() * (max - min);
  }
}
