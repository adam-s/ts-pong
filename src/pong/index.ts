export * from './Pong';
export * from './Ball';
export * from './Court';
export * from './Menu';
export * from './Paddle';
export * from './resources';
