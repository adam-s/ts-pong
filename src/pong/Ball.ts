/* eslint-disable @typescript-eslint/no-empty-function */
import { Engine } from '../engine';
import { accelerate, Drawable, intercept, Updatable } from '../utils';
import { Paddle } from '../pong';
import { Colors } from './Pong';

export interface BallOptions {
  ballRadius: number;
  width: number;
  height: number;
  wallWidth: number;
  ballSpeed: number;
  ballAccel: number;
  footprints: boolean;
}

export class Ball implements Updatable, Drawable {
  private _x = 0;

  private _y = 0;

  private _dx: number;

  private _dy: number;

  private _radius: number;

  private minX: number;

  private maxX: number;

  private minY: number;

  private maxY: number;

  private speed: number;

  private accel: number;

  // private top: number;

  // private bottom: number;

  private _left: number;

  private _right: number;

  private dxChanged = false;

  private dyChanged = false;

  private footprints: Array<{ x: number; y: number }> = [];

  private footprintCount = 0;

  public showFootprints = false;

  constructor(options: BallOptions) {
    const {
      ballRadius,
      width,
      height,
      wallWidth,
      ballSpeed,
      ballAccel,
      footprints,
    } = options;
    this.minX = this._radius = ballRadius;
    this.maxX = width - this._radius;
    this.minY = wallWidth + this._radius;
    this.maxY = height - wallWidth - this._radius;
    this.speed = (this.maxX - this.minX) / ballSpeed;
    this.accel = ballAccel;
    this.showFootprints = footprints;
  }

  get x() {
    return this._x;
  }

  get y() {
    return this._y;
  }

  get dx() {
    return this._dx;
  }

  get dy() {
    return this._dy;
  }

  get radius() {
    return this._radius;
  }

  get left() {
    return this._left;
  }

  get right() {
    return this._right;
  }

  reset(playerNumber: number | null) {
    this.footprints = [];
    this.setpos(
      playerNumber === 1 ? this.maxX : this.minX,
      Engine.random(this.minY, this.maxY)
    );
    this.setdir(playerNumber === 1 ? -this.speed : this.speed, this.speed);
  }

  setpos(x: number, y: number) {
    this._x = x;
    this._y = y;
    this._left = this._x - this.radius;
    // this.top = this._y - this.radius;
    this._right = this._x + this.radius;
    // this.bottom = this._y + this.radius;
  }

  setdir(dx: number, dy: number) {
    this.dxChanged = this._dx < 0 !== dx < 0; // did horizontal direction change
    this.dyChanged = this._dy < 0 !== dy < 0; // did vertical direction change
    this._dx = dx;
    this._dy = dy;
  }

  footprint() {
    if (this.showFootprints) {
      if (!this.footprintCount || this.dxChanged || this.dyChanged) {
        this.footprints.push({ x: this.x, y: this.y });
        if (this.footprints.length > 50) this.footprints.shift();
        this.footprintCount = 5;
      } else {
        this.footprintCount--;
      }
    }
  }

  intercept(
    rect: { right: number; left: number; top: number; bottom: number },
    nx: number,
    ny: number
  ) {
    let pt;
    // Moving left
    if (nx < 0) {
      pt = intercept({
        x1: this.x, // x coord on the right side of ball segment
        y1: this.y, // y coord on the right side of ball segment
        x2: this.x + nx, // x coord where ball will be painted moving left
        y2: this.y + ny, // y coord where the ball will be painted moving left
        x3: rect.right + this.radius,
        y3: rect.top - this.radius,
        x4: rect.right + this.radius,
        y4: rect.bottom + this.radius,
        d: 'right',
      });
    } else if (nx > 0) {
      pt = intercept({
        x1: this.x,
        y1: this.y,
        x2: this.x + nx,
        y2: this.y + ny,
        x3: rect.left - this.radius,
        y3: rect.top - this.radius,
        x4: rect.left - this.radius,
        y4: rect.bottom + this.radius,
        d: 'left',
      });
    }
    if (!pt) {
      if (ny < 0) {
        pt = intercept({
          x1: this.x,
          y1: this.y,
          x2: this.x + nx,
          y2: this.y + ny,
          x3: rect.left - this.radius,
          y3: rect.bottom + this.radius,
          x4: rect.right + this.radius,
          y4: rect.bottom + this.radius,
          d: 'bottom',
        });
      } else if (ny > 0) {
        pt = intercept({
          x1: this.x,
          y1: this.y,
          x2: this.x + nx,
          y2: this.y + ny,
          x3: rect.left - this.radius,
          y3: rect.top - this.radius,
          x4: rect.right + this.radius,
          y4: rect.top - this.radius,
          d: 'top',
        });
      }
    }
    return pt;
  }

  update(dt: number, leftPaddle: Paddle, rightPaddle: Paddle) {
    const pos = accelerate({
      x: this.x,
      y: this.y,
      dx: this.dx,
      dy: this.dy,
      accel: this.accel,
      dt,
    });

    // Reverse direction (up) if ball hits bottom wall
    if (pos.dy > 0 && pos.y > this.maxY) {
      pos.y = this.maxY; // Set the y position to the top of the bottom wal
      pos.dy = -pos.dy; // Reverse direction
      // Reverse direction (down) if ball hits top wall
    } else if (pos.dy < 0 && pos.y < this.minY) {
      pos.y = this.minY; // Set the y position to the bottom of the top wall
      pos.dy = -pos.dy; // Reverse direction
    }

    // Choose paddle based on direction of ball. Negative from right to left is left paddle
    const paddle = pos.dx < 0 ? leftPaddle : rightPaddle;
    // Test if the ball will intercept the current paddle
    const pt = this.intercept(paddle, pos.nx, pos.ny);
    // If there is an interception with a paddle set the position to the point of intercept
    // Switch directions based on left / right or top / bottom
    if (pt) {
      switch (pt.d) {
        case 'left':
        case 'right':
          pos.x = pt.x;
          pos.dx = -pos.dx;
          break;
        case 'top':
        case 'bottom':
          pos.y = pt.y;
          pos.dy = -pos.dy;
          break;
      }

      // add/remove spin based on paddle direction
      if (paddle.up) pos.dy = pos.dy * (pos.dy < 0 ? 0.5 : 1.5);
      else if (paddle.down) pos.dy = pos.dy * (pos.dy > 0 ? 0.5 : 1.5);
    }

    this.setpos(pos.x, pos.y);
    this.setdir(pos.dx, pos.dy);
    this.footprint();
  }

  draw(ctx: CanvasRenderingContext2D) {
    const side = this.radius * 2;
    ctx.fillStyle = Colors.ball;
    ctx.fillRect(this.x - this.radius, this.y - this.radius, side, side);
    if (this.showFootprints) {
      const max = this.footprints.length;
      ctx.strokeStyle = Colors.footprint;
      for (let n = 0; n < max; n++)
        ctx.strokeRect(
          this.footprints[n].x - this.radius,
          this.footprints[n].y - this.radius,
          side,
          side
        );
    }
  }
}
