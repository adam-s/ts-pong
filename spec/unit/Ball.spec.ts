/* eslint-disable @typescript-eslint/no-var-requires */
import 'jest-canvas-mock';
import { Ball } from '../../src/pong/Ball';
import { Engine } from '../../src/engine/Engine';
import * as utils from '../../src/utils/utils';
import { Paddle } from '../../src/pong/Paddle';
jest.mock('../../src/engine/Engine');

describe('Ball', () => {
  it('should exist', () => {
    expect(Ball).toBeDefined();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
  });

  describe('instance', () => {
    let implementation: Ball;
    const Config = require('../../src/pong/Pong').Config;

    beforeEach(() => {
      implementation = new Ball(Config);
    });

    it('instantiates a ball', () => {
      expect(implementation).toMatchSnapshot();
    });

    it('resets', () => {
      // https://stackoverflow.com/questions/50421732/mocking-up-static-methods-in-jest
      Engine.random = jest.fn().mockReturnValue(1);
      implementation.setdir = jest.fn();
      implementation.setpos = jest.fn();
      implementation.reset(1);
      expect(implementation['footprints']).toEqual([]);
      expect(implementation.setdir).toBeCalledWith(-157.5, 157.5);
      expect(implementation.setpos).toBeCalledWith(635, 1);
    });

    it('setpos', () => {
      implementation.setpos(10, 10);
      expect(implementation).toMatchSnapshot();
    });

    it('setdir', () => {
      implementation.setdir(-191, 1);
      expect(implementation).toMatchSnapshot();
    });
  });

  describe('footprints', () => {
    jest.mock('../../src/pong/Pong', () => {
      const originalModule = jest.requireActual('../../src/pong/Pong');
      return {
        __esModule: true,
        ...originalModule,
        Config: {
          ...originalModule.Config,
          footprints: true,
        },
      };
    });

    const Config = require('../../src/pong/Pong').Config;
    let ball: Ball;
    beforeEach(() => {
      ball = new Ball(Config);
    });

    it('should create footprints', () => {
      ball.footprint();
      expect(ball['footprints']).toEqual([{ x: 0, y: 0 }]);
    });

    it('should decrement footprintCount by 1', () => {
      ball.footprint();
      expect(ball['footprintCount']).toBe(5);
      ball.footprint();
      expect(ball['footprintCount']).toBe(4);
      ball.footprint();
      expect(ball['footprintCount']).toBe(3);
      ball.footprint();
      expect(ball['footprintCount']).toBe(2);
      ball.footprint();
      expect(ball['footprintCount']).toBe(1);
      ball.footprint();
      expect(ball['footprintCount']).toBe(0);
      ball.footprint();
      expect(ball['footprintCount']).toBe(5);
    });
  });
  // Intercept
  describe('intercept', () => {
    let implementation: Ball;
    const Config = require('../../src/pong/Pong').Config;

    beforeEach(() => {
      implementation = new Ball(Config);
    });

    /**
     * x1 621.0463999999953
     * y1 42.18688000000225
     * x2 624.0263039999953
     * y2 39.20902400000228
     * x3 623
     * y3 10.168
     * x4 623
     * y4 80.168
     * d left
     * ball.x 621.0463999999953
     * ball.y 42.18688000000225
     * nx 2.979904000000033
     * ny -2.977855999999967
     * rect.top 15.168
     * rect.left 628
     * rect.bottom 75.168
     * ball.radius 5
     * d: "left"
     * x: 623
     * y: 40.234622651572934
     */
    it('call intercept with left', () => {
      const interceptSpy = jest.spyOn(utils, 'intercept').mockReturnValue({
        d: 'left',
        x: 623,
        y: 40.234622651572934,
      });
      const rect = { right: 100, left: 628, top: 15.168, bottom: 75.168 };
      const nx = 2.979904000000033;
      const ny = -2.977855999999967;
      implementation['_x'] = 621.0463999999953;
      implementation['_y'] = 42.18688000000225;
      implementation.intercept(rect, nx, ny);
      expect(interceptSpy.mock.calls[0][0]).toEqual({
        x1: 621.0463999999953,
        y1: 42.18688000000225,
        x2: 624.0263039999953,
        y2: 39.20902400000228,
        x3: 623,
        y3: 10.168,
        x4: 623,
        y4: 80.168,
        d: 'left',
      });
      interceptSpy.mockClear();
    });
    /**
     * x1 19.626584000007547
     * y1 329.49776800000086
     * x2 16.257304000007604
     * y2 326.1284880000009
     * x3 17
     * y3 309.34600000000023
     * x4 17
     * y4 379.34600000000023
     * d right
     * ball.x 19.626584000007547
     * ball.y 329.49776800000086
     * nx -3.369279999999943
     * ny -3.3692799999999465
     * rect.right 12
     * rect.top 314.34600000000023
     * rect.bottom 374.34600000000023
     * ball.radius 5
     * pt {x: 17, y: 326.8711839999933, d: 'right'}
     */
    it('call intercept with right', () => {
      const interceptSpy = jest
        .spyOn(utils, 'intercept')
        .mockReturnValue({ x: 17, y: 326.8711839999933, d: 'right' });
      const rect = {
        right: 12,
        left: 0,
        top: 314.34600000000023,
        bottom: 374.34600000000023,
      };
      const nx = -3.369279999999943;
      const ny = -3.369279999999943;
      implementation['_x'] = 19.626584000007547;
      implementation['_y'] = 329.49776800000086;
      implementation.intercept(rect, nx, ny);
      expect(interceptSpy.mock.calls[0][0]).toEqual({
        x1: 19.626584000007547,
        y1: 329.49776800000086,
        x2: 16.257304000007604,
        y2: 326.1284880000009,
        x3: 17,
        y3: 309.34600000000023,
        x4: 17,
        y4: 379.34600000000023,
        d: 'right',
      });
      interceptSpy.mockClear();
    });
    /**
     *  x1 620.5650559999834
     *  y1 17
     *  x2 624.6336559999834
     *  y2 22.964209999999852
     *  x3 623
     *  y3 22.444000000000017
     *  x4 645
     *  y4 22.444000000000017
     *  d top
     *  ball.x 620.5650559999834
     *  ball.y 17
     *  nx 4.068599999999947
     *  ny 5.964209999999852
     *  rect.left 628
     *  rect.top 27.444000000000017
     *  rect.right 640
     *  ball.radius 5
     *  pt {x: 624.2787847922292, y: 22.444000000000017, d: 'top'}
     */
    it('call intercept with top', () => {
      const interceptSpy = jest
        .spyOn(utils, 'intercept')
        .mockReturnValueOnce(null)
        .mockReturnValue({
          x: 624.2787847922292,
          y: 22.444000000000017,
          d: 'top',
        });
      const rect = {
        left: 628,
        top: 27.444000000000017,
        right: 640,
        bottom: 0,
      };
      const nx = 4.068599999999947;
      const ny = 5.964209999999852;
      implementation['_x'] = 620.5650559999834;
      implementation['_y'] = 17;
      implementation.intercept(rect, nx, ny);
      expect(interceptSpy.mock.calls[1][0]).toEqual({
        x1: 620.5650559999834,
        y1: 17,
        x2: 624.6336559999834,
        y2: 22.964209999999852,
        x3: 623,
        y3: 22.444000000000017,
        x4: 645,
        y4: 22.444000000000017,
        d: 'top',
      });
      interceptSpy.mockClear();
    });
    /**
     *  x1 620.9805519999878
     *  y1 463
     *  x2 626.3485519999878
     *  y2 455.06838400000015
     *  x3 623
     *  y3 455.7740000000003
     *  x4 645
     *  y4 455.7740000000003
     *  d bottom
     *  ball.x 620.9805519999878
     *  ball.y 463
     *  nx 5.367999999999938
     *  ny -7.931615999999849
     *  rect.left 628
     *  rect.bottom 450.7740000000003
     *  rect.right 640
     *  ball.radius 5
     *  pt {x: 625.871001562851, y: 455.7740000000003, d: 'bottom'}
     */
    it('call intercept with bottom', () => {
      const interceptSpy = jest
        .spyOn(utils, 'intercept')
        .mockReturnValueOnce(null)
        .mockReturnValue({
          x: 625.871001562851,
          y: 455.7740000000003,
          d: 'bottom',
        });
      const rect = {
        left: 628,
        bottom: 450.7740000000003,
        right: 640,
        top: 0,
      };
      const nx = 5.367999999999938;
      const ny = -7.931615999999849;
      implementation['_x'] = 620.9805519999878;
      implementation['_y'] = 463;
      implementation.intercept(rect, nx, ny);
      expect(interceptSpy.mock.calls[1][0]).toEqual({
        x1: 620.9805519999878,
        y1: 463,
        x2: 626.3485519999878,
        y2: 455.06838400000015,
        x3: 623,
        y3: 455.7740000000003,
        x4: 645,
        y4: 455.7740000000003,
        d: 'bottom',
      });
      interceptSpy.mockClear();
    });
  });
  describe('update', () => {
    const Config = require('../../src/pong/Pong').Config;

    let implementation: Ball;
    let setposSpy: jest.SpyInstance;
    let setdirSpy: jest.SpyInstance;
    let footprintSpy: jest.SpyInstance;
    let interceptSpy: jest.SpyInstance;

    const accelerateSpy = jest.spyOn(utils, 'accelerate');

    let leftPaddle: Paddle;
    let rightPaddle: Paddle;

    beforeEach(() => {
      implementation = new Ball(Config);
      setposSpy = jest
        .spyOn(implementation, 'setpos')
        .mockImplementation(() => undefined);
      setdirSpy = jest
        .spyOn(implementation, 'setdir')
        .mockImplementation(() => undefined);
      footprintSpy = jest
        .spyOn(implementation, 'footprint')
        .mockImplementation(() => undefined);
      interceptSpy = jest
        .spyOn(implementation, 'intercept')
        .mockImplementation(() => undefined);

      leftPaddle = new Paddle(Config);
      rightPaddle = new Paddle(Config);
      jest.clearAllMocks();
    });

    it('hits bottom wall then changes direction up', () => {
      accelerateSpy.mockImplementation(() => ({
        dx: 168.13199999999938,
        dy: 168.13199999999938,
        nx: 2.8570880000000045,
        ny: 2.857087999999976,
        x: 221.38246399999957,
        y: 465.7664818886045,
      }));
      implementation.update(0, leftPaddle, rightPaddle);
      expect(setposSpy).toBeCalledWith(221.38246399999957, 463);
      expect(setdirSpy).toBeCalledWith(168.13199999999938, -168.13199999999938);
      expect(footprintSpy).toBeCalled();
    });

    it('hits top wall then changes direction down', () => {
      accelerateSpy.mockImplementation(() => ({
        dx: -189.68399999999804,
        dy: -189.68399999999804,
        nx: -3.0318719999999075,
        ny: -3.0318719999999697,
        x: 544.4649520000016,
        y: 14.450568000001619,
      }));
      implementation.update(0, leftPaddle, rightPaddle);
      expect(setposSpy).toBeCalledWith(544.4649520000016, 17);
      expect(setdirSpy).toBeCalledWith(-189.68399999999804, 189.68399999999804);
      expect(footprintSpy).toBeCalled();
    });

    it('hits left / right paddle then changes direction', () => {
      accelerateSpy.mockImplementation(() => ({
        dx: 186.2839999999989,
        dy: -186.2839999999989,
        nx: 3.3518159999999853,
        ny: -3.3492239999999818,
        x: 623.4674159999983,
        y: 27.49023200000089,
      }));
      interceptSpy.mockImplementation(() => ({
        d: 'left',
        x: 623,
        y: 27.95728654153245,
      }));
      implementation.update(0, leftPaddle, rightPaddle);
      expect(setposSpy).toBeCalledWith(623, 27.49023200000089);
      expect(setdirSpy).toBeCalledWith(-186.2839999999989, -186.2839999999989);
      expect(footprintSpy).toBeCalled();
    });

    it('hits top / bottom paddle then changes direction', () => {
      accelerateSpy.mockImplementation(() => ({
        dx: 363.5799999999966,
        dy: -312.63499999999544,
        nx: 5.452799999999911,
        ny: -4.686824999999942,
        x: 633.6720559999937,
        y: 178.57399999999942,
      }));
      interceptSpy.mockImplementation(() => ({
        x: 631.2019512412338,
        y: 178.57399999999942,
        d: 'bottom',
      }));
      implementation.update(0, leftPaddle, rightPaddle);
      expect(setposSpy).toBeCalledWith(633.6720559999937, 178.57399999999942);
      expect(setdirSpy).toBeCalledWith(363.5799999999966, 312.63499999999544);
      expect(footprintSpy).toBeCalled();
    });

    it('puts spin on the ball when paddle moves down', () => {
      accelerateSpy.mockImplementation(() => ({
        dx: 210.74799999999772,
        dy: -210.74799999999772,
        nx: 3.5815599999999677,
        ny: -3.579248000000007,
        x: 623.8756399999951,
        y: 452.47927200000004,
      }));
      interceptSpy.mockImplementation(() => ({
        x: 623,
        y: 453.3543467491882,
        d: 'left',
      }));
      rightPaddle['_down'] = 1;
      implementation.update(0, leftPaddle, rightPaddle);
      expect(setposSpy).toBeCalledWith(623, 452.47927200000004);
      expect(setdirSpy).toBeCalledWith(
        -210.74799999999772,
        -316.12199999999655
      );
      expect(footprintSpy).toBeCalled();
    });
  });
  describe('draw', () => {
    let ctx: CanvasRenderingContext2D;
    const Config = require('../../src/pong/Pong').Config;
    let implementation: Ball;

    const canvas = document.createElement('canvas');
    ctx = canvas.getContext('2d');
    beforeEach(() => {
      const canvas = document.createElement('canvas');
      ctx = canvas.getContext('2d');
      implementation = new Ball(Config);
    });

    afterEach(() => {
      ctx.__clearEvents();
    });

    it('renders ball to canvas', () => {
      implementation.draw(ctx);
      const events = ctx.__getEvents();
      expect(events).toMatchSnapshot();
    });

    it('renders footprints to canvas', () => {
      const footprints = [
        { x: 258.9049039999994, y: 303.8599039999993 },
        { x: 275.25293599999924, y: 287.52415999999914 },
        { x: 291.674695999999, y: 271.1147039999991 },
        { x: 308.5146239999988, y: 254.2875919999991 },
        { x: 324.73939999999857, y: 238.07462399999923 },
        { x: 341.38234399999834, y: 221.44396799999933 },
        { x: 357.9245039999982, y: 204.91386399999945 },
        { x: 374.88941599999794, y: 187.9615519999996 },
        { x: 391.75354399999776, y: 171.10975999999977 },
        { x: 408.69139999999754, y: 154.1842079999999 },
      ];
      implementation['footprints'] = footprints;
      implementation['showFootprints'] = true;
      implementation.draw(ctx);
      const events = ctx.__getEvents();
      expect(events).toMatchSnapshot();
    });
  });
});
