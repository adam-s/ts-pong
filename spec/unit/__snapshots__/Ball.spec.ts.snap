// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ball draw renders ball to canvas 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": -5,
      "y": -5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#333333",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
]
`;

exports[`Ball draw renders footprints to canvas 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": -5,
      "y": -5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#333333",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 253.90490399999942,
      "y": 298.8599039999993,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 270.25293599999924,
      "y": 282.52415999999914,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 286.674695999999,
      "y": 266.1147039999991,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 303.5146239999988,
      "y": 249.2875919999991,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 319.73939999999857,
      "y": 233.07462399999923,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 336.38234399999834,
      "y": 216.44396799999933,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 352.9245039999982,
      "y": 199.91386399999945,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 369.88941599999794,
      "y": 182.9615519999996,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 386.75354399999776,
      "y": 166.10975999999977,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 10,
      "width": 10,
      "x": 403.69139999999754,
      "y": 149.1842079999999,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
]
`;

exports[`Ball instance instantiates a ball 1`] = `
Ball {
  "_radius": 5,
  "_x": 0,
  "_y": 0,
  "accel": 8,
  "dxChanged": false,
  "dyChanged": false,
  "footprintCount": 0,
  "footprints": Array [],
  "maxX": 635,
  "maxY": 463,
  "minX": 5,
  "minY": 17,
  "showFootprints": undefined,
  "speed": 157.5,
}
`;

exports[`Ball instance setdir 1`] = `
Ball {
  "_dx": -191,
  "_dy": 1,
  "_radius": 5,
  "_x": 0,
  "_y": 0,
  "accel": 8,
  "dxChanged": true,
  "dyChanged": false,
  "footprintCount": 0,
  "footprints": Array [],
  "maxX": 635,
  "maxY": 463,
  "minX": 5,
  "minY": 17,
  "showFootprints": undefined,
  "speed": 157.5,
}
`;

exports[`Ball instance setpos 1`] = `
Ball {
  "_left": 5,
  "_radius": 5,
  "_right": 15,
  "_x": 10,
  "_y": 10,
  "accel": 8,
  "bottom": 15,
  "dxChanged": false,
  "dyChanged": false,
  "footprintCount": 0,
  "footprints": Array [],
  "maxX": 635,
  "maxY": 463,
  "minX": 5,
  "minY": 17,
  "showFootprints": undefined,
  "speed": 157.5,
  "top": 5,
}
`;
